python/arcticdb/storage_fixtures/mongo.py

```diff
@@ -0,0 +1,145 @@
+
+    def set_permission(self, *, read: bool, write: bool):
+        raise NotImplementedError("Will support setting permissions on Mongo soon")  # TODO
+
+
+class ExternalMongoDBServer(StorageFixtureFactory):
+    """A MongoDB server whose life-cycle is managed externally to this test fixture system."""
+
+    def __init__(self, mongo_uri: str):
+        self.mongo_uri = mongo_uri
+
+    def __str__(self):
+        return f"{type(self).__name__}[{self.mongo_uri}]"
+
+    def create_fixture(self) -> StorageFixture:
+        return MongoDatabase(self.mongo_uri)
+
+
+class ManagedMongoDBServer(StorageFixtureFactory):
+    """Represents a MongoDB server started by this class"""
+
+    _count = -1
+
+    def __init__(self, data_dir: Optional[str] = None, port=0, executable="mongod"):
+        self._data_dir = data_dir or tempfile.mkdtemp("ManagedMongoDBServer")
+        self._port = port or get_ephemeral_port()
+        self._executable = executable
+
+    def _safe_enter(self):
+        from pymongo import MongoClient
+
+        cmd = [self._executable, "--port", str(self._port), "--dbpath", self._data_dir, "--nojournal"]
+        self._p = GracefulProcessUtils.start(cmd)
+        self.mongo_uri = f"mongodb://localhost:{self._port}"
+        wait_for_server_to_come_up(f"http://localhost:{self._port}", "mongod", self._p)
+        self._client = MongoClient(self.mongo_uri)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self._client:
+            with handle_cleanup_exception(self):
+                self._client["admin"].command({"shutdown": 1, "force": True, "timeoutSecs": 1})
+
+        with handle_cleanup_exception(self, "process", consequence="On-disk data might not be delete-able. "):
+            GracefulProcessUtils.terminate(self._p)
+
+        try:
+            shutil.rmtree(self._data_dir, ignore_errors=False)
+        except:
+            time.sleep(1)
+            # Even with ignore_errors=True, rmtree might still throw on Windows....
+            with handle_cleanup_exception(self, "files (best effort)", consequence="Disk might get used up"):
+                shutil.rmtree(self._data_dir, ignore_errors=True)
+
+    def create_fixture(self) -> StorageFixture:
+        self._count += 1
+        return MongoDatabase(self.mongo_uri, f"Managed{self._count}")
+
+    def __str__(self):
+        return f"{type(self).__name__}[{self.mongo_uri}]"
+
+
+def auto_detect_server():
+    """Use the Server specified by the CI_MONGO_HOST env var or localhost, if available, falling back to starting a
+    dedicated instance on a random port."""
+    import requests
+
+    mongo_host = os.getenv("CI_MONGO_HOST", "localhost")
+    try:
+        res = requests.get(f"http://{mongo_host}:27017")
+        if res.status_code == 200 and "mongodb" in res.text.lower():
+            return ExternalMongoDBServer(f"mongodb://{mongo_host}:27017")
+    except requests.exceptions.ConnectionError:
+        return ManagedMongoDBServer()
```

Comment:
Is the fallback desirable? I would think that if we have `CI_MONGO_HOST` specified it would be surprising behaviour to silently _not_ use it!

Reply:
Good point. Will change

Reply:
I'll make the change, but it won't be able to support the `CI_MONGO_HOST` defaults to `localhost` behaviour any more.

In the medium term, we can install the `mongod` binary into our Docker image (it's already installed in other runners) and then drop the `CI_MONGO_HOST`.